packages/core/test/after.test.ts

```diff
@@ -85,77 +85,83 @@ describe('delayed transitions', () => {
     `);
   });
 
-  it('should be able to transition with delay from nested initial state', (done) => {
-    const machine = createMachine({
-      initial: 'nested',
-      states: {
-        nested: {
-          initial: 'wait',
-          states: {
-            wait: {
-              after: {
-                10: '#end'
+  it('should be able to transition with delay from nested initial state', () =>
+    new Promise<void>((resolve) => {
```

Comment:
Those kind of changes could probably be replaced with `async` functions and `waitFor`. Some might not be but then I'd go with `Promise.withResolvers` (which might not be available in our testing envs, I'm not sure). This would allow u to create a deferred thing at the top of the async function, await it at the bottom before ur `expect` calls and `.resolve`/`.reject` it from callbacks. This would avoid adding extra indentation to the tests that got refactored by introducing `new Promise`

Reply:
imo the `new Promise` callback style works well for assertion-free tests (where the test is basically just that it doesn't time out (because `resolve` was called)).

I think this example shows the pitfalls of `waitFor`/`Promise.withResolvers` with this style of test:

```ts
test('new Promise', () =>
  new Promise((resolve) => {
    const machine = createMachine({
      initial: 'waiting',
      states: {
        waiting: {
          after: {
            10: 'done'
          }
        },
        done: {
          entry: resolve
        }
      }
    });
    createActor(machine).start();

    // no assertions or other logic necessary. the test only passes if the
    // promise resolves; otherwise, it times out
  }));

test('waitFor', async () => {
  const machine = createMachine({
    initial: 'waiting',
    states: {
      waiting: {
        after: {
          10: 'done'
        }
      },
      done: {}
    }
  });
  const actor = createActor(machine);
  actor.start();

  // we need to wait for + assert the transition. if we forget to `await` or
  // forget `waitFor` entirely, the test passes even though we haven't actually
  // tested anything
  await waitFor(actor, (snapshot) => snapshot.matches('done'));
});

test('Promise.withResolvers', async () => {
  const { promise, resolve } = Promise.withResolvers();

  const machine = createMachine({
    initial: 'waiting',
    states: {
      waiting: {
        after: {
          10: 'done'
        }
      },
      done: {
        entry: resolve
      }
    }
  });
  const actor = createActor(machine);
  actor.start();

  // if we don't `await promise` or `return promise`, the test passes even
  // though `resolve` isn't called. would be an easy mistake to make and the
  // cost is that the test isn't actually testing anything even though it's
  // passing
  return promise;
});
```

Definitely agree with `waitFor`/`Promise.withResolvers` for tests where we're actually making assertions and such though.

Reply:
> Though actually now that I'm thinking about it, maybe we shouldn't have assertion-free tests to begin with 🙃

This. I think it's always best to assert something explicitly. That always gives some extra insight to what is actually tested